cumulusci/core/tasks.py

```diff
@@ -226,6 +236,115 @@ def freeze(self, step):
+    Python stack.
+
+    Attributes of resumable tasks are persisted by default.
+
+    __setattr__ is overriden to require you to declare values and types
+    in order to encourage you to think through what should be
+    persisted and what should not.
+
+    Add transient attributes through self.set_nonpersistent_value to
+    prevent them from being persisted.
+    """
+
+    options: dict = {}
+    name: str = None
+    logger: Any = None
+    resumption_fs_resource: Any = None
+    project_config: dict = None
+    task_config: dict = None
+    org_config: dict = None
+    poll_count: int = None
+    poll_interval_level: int = None
+    poll_interval_s: int = None
+    poll_complete: bool = None
+    return_values: dict = None
+    result: Any = None
+    flow: Any = None
+    stepnum: Any = None
+
+    default_excluded_fields = (
+        "name",
+        "logger",
+        "task_options",
+        "project_config",
+        "task_config",
+        "org_config",
+        "poll_count",
+        "poll_interval_level",
+        "poll_interval_s",
+        "poll_complete",
+        "return_values",
+        "result",
+        "flow",
+        "stepnum",
+        "excluded_fields",
+        "working_path",
+        "sf",
+        "bulk",
+        "tooling",
+    )
+
+    def __init__(self, *args, **kwargs):
+        self.__dict__["excluded_fields"] = set(self.default_excluded_fields)
+
+        if "resumption_fs_resource" in kwargs:
+            self.set_nonpersistent_value(
+                "resumption_fs_resource", kwargs.pop("resumption_fs_resource")
+            )
+
+        super().__init__(*args, **kwargs)
+
+    def _run_task(self):
+        self._start()
+        self._run_steps()
+        self.resumption_fs_resource.cleanup()
+        self._after_finished()
+
+    def _start(self):
+        "Override this: Do any kind of task setup or logging that is necessary."
+        self.logger.info(f"Starting task {self.__class__.name}")
+
+    def _run_steps(self):
+        "Run multiple steps until finished. Usually do not override this."
+        while not self._is_finished():
+            self._run_step()
+            self.resumption_fs_resource.state_data = self.dict()
+            self.resumption_fs_resource.save()
+
+    def resume(self):
+        "Resume this task. Usually do not need to override this."
+        self._resume()
+        self._run_steps()
+        self._after_finished()
+
+    def set_nonpersistent_value(self, key, value):
+        "Save a value to memory but not to persistent storage"
+        self.__dict__[key] = value
+        self.excluded_fields.add(key)
+
+    def dict(self):
+        return {k: v for k, v in self.__dict__.items() if k not in self.excluded_fields}
+
+    def __setattr__(self, name, value):
+        if name not in self.excluded_fields and not hasattr(self.__class__, name):
+            raise AttributeError(f"Undeclared attribute {name}")
+        self.__dict__[name] = value
```

Comment:
My basic feeling is that almost all state should be in the resumption file or on the stack. Neither of the first two tasks converted to this model use set_nonpersistent_value at all.

So the fact that it is inconvenient is fine because it is rare, and hopefully the fact that it is inconvenient will help ensure it IS rare. I think that mixing a lot of persistent and non-persistent data is just likely to confuse the programmer. The base class uses that function a few times because its just the nature of frameworks that they use a bit more magic than the stuff built on top of them.

Reply:
Okay, I am on board with the argument for making it a bit more work to store nonpersistent state in instance attributes. (It helps that there are a number of existing tasks where I think instance attributes are overused.)

But you didn't answer the other part of my question, about the possibility of keeping persistent state more explicitly contained in a separate `state` object. My arguments for this are:
- It makes it clearer which values are being persisted; instead of having to keep track of which names are being stored which way, you can just look at where the value is being set.
- It would remove the need to manage a list of excluded fields for each task, and clean up the ResumableTask implementation a bit since it wouldn't need to filter based on what's in excluded_fields
- It makes resuming more straightforward: The task can work with a state object reconstructed in its entirety from the resumption file, instead of needing to transfer specific state values to specific task attributes.

Reply:
@prescod copying your comment here since github didn't attach it to the right thread.

> Do you mean that the programmer would need to refer to all persistent state
as `self.state.xyz`

yes, this one. explicit is better than implicit / namespaces are one honking great idea

> Of do you mean that the base-class would override setattr/getattr to
transparently store its state elsewhere?

No.

Reply:
I actually implemented it that way at first, but then it seemed like a bunch of extra typing to no benefit for the task creator.

```
self.state.this = 
self.state.that =
```

Plus two classes per task (the task and its strong-typed state). And two base classes to import (the BaseResumableTask and its BaseResumptionState).

It seemed like a lot of boilerplate for not much (if any) benefit (for the Task programmer) so I combined them.

Reply:
I still disagree. To me, it feels like a worthwhile separation of concerns.